File: api/code/api.fsharp/Azure.fs
[<RequireQualifiedAccess>]
module Azure

open Azure.Core
open Azure.Identity
open Microsoft.Extensions.Configuration
open FSharpPlus
open common
open Microsoft.Extensions.Hosting

let private getAuthorityHost provider =
    ServiceProvider.getService<IConfiguration> provider
    |> bind (fun configuration -> Configuration.getValue configuration "AZURE_CLOUD_ENVIRONMENT")
    |> (function
    | Some "AzureUSGovernment"
    | Some "AzureGovernment" -> AzureAuthorityHosts.AzureGovernment
    | Some "AzureChina" -> AzureAuthorityHosts.AzureChina
    | Some "AzureGlobalCloud"
    | Some "AzureCloud"
    | Some "AzurePublicCloud" -> AzureAuthorityHosts.AzurePublicCloud
    | Some value ->
        failwith
            $"'{value}' is not valid for 'AZURE_CLOUD_ENVIRONMENT'. Allowed values are 'AzureGovernment', 'AzureChina', and 'AzurePublicCloud'."
    | None -> AzureAuthorityHosts.AzurePublicCloud)

let private getTokenCredential provider : TokenCredential =
    let authorityHost = getAuthorityHost provider

    let options = DefaultAzureCredentialOptions()
    options.AuthorityHost <- authorityHost
    options.ExcludeVisualStudioCredential <- true
    options.ExcludeVisualStudioCodeCredential <- true

    DefaultAzureCredential(options)

let configureTokenCredential (builder: IHostApplicationBuilder) =
    ServiceCollection.tryAddSingleton builder.Services getTokenCredential

File: common/code/common/Azure.cs
using Azure.Core;
using Azure.Identity;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection.Extensions;
using Microsoft.Extensions.Hosting;
using System;

namespace common;

public static class AzureModule
{
    public static void ConfigureTokenCredential(IHostApplicationBuilder builder) =>
        builder.Services.TryAddSingleton(GetTokenCredential);

    private static TokenCredential GetTokenCredential(IServiceProvider provider)
    {
        var authorityHost = GetAuthorityHost(provider);

        var options = new DefaultAzureCredentialOptions
        {
            AuthorityHost = authorityHost,
            ExcludeVisualStudioCodeCredential = true,
            ExcludeVisualStudioCredential = true
        };

        return new DefaultAzureCredential(options);
    }

    private static Uri GetAuthorityHost(IServiceProvider provider) =>
        provider.GetRequiredService<IConfiguration>()
                .GetValue("AZURE_CLOUD_ENVIRONMENT")
                .Map(environment => environment switch
                {
                    "AzureGlobalCloud" or nameof(AzureAuthorityHosts.AzurePublicCloud) =>
                        AzureAuthorityHosts.AzurePublicCloud,
                    "AzureUSGovernment" or nameof(AzureAuthorityHosts.AzureGovernment) =>
                        AzureAuthorityHosts.AzureGovernment,
                    nameof(AzureAuthorityHosts.AzureChina) => AzureAuthorityHosts.AzureChina,
                    _ => throw new InvalidOperationException($"""
'{environment}' is not a valid cloud environment. Valid values are {nameof(AzureAuthorityHosts.AzurePublicCloud)}, {nameof(AzureAuthorityHosts.AzureChina)}, {nameof(AzureAuthorityHosts.AzureGovernment)}.
""")
                })
                .IfNone(() => AzureAuthorityHosts.AzurePublicCloud);
}